### ISSUE TITLE
[cfe] Wrong context type for map pattern declaration?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#51414]
Consider the following program:

```dart
X whatever<X>() {
  print(X);
  return <Never>[] as X;
}

void main() {
  var ({"key2": List<int?> xs, ...} && map) = {"key": [], "key2": whatever()};
  // int _ = map; // Analyzer type for map is `Map<String, List<int?>>`.
}
```

This program is accepted by the analyzer (and if we uncomment the last line in `main` then we can see that the static type of `map` is considered to be `Map<String, List<int?>>`).

However, if that same line is uncommented then the CFE (version 3.0.0-204.0.dev) reports the type `Map<String, dynamic>` for `map`, which seems to indicate that it does not use the context type as specified [here](https://github.com/dart-lang/language/blob/master/accepted/future-releases/0546-patterns/feature-specification.md#:~:text=Else%20K%20is%20_%20and%20V%20is%20the%20greatest%20lower%20bound).

It should be the greatest lower bound of the singleton set containing `List<int?>`, which is again `List<int?>`.

When the last line of `main` is commented out it is possible to execute the program, and it then prints `dynamic` (which shows that the value `whatever()` is inferred with the context type `dynamic`, which is consistent with the type of the map as a whole being `Map<String, dynamic>`).

In other words, the only discrepancy is that the CFE seems to forget to use the type obtained from the values of the map pattern when it computes the pattern type schema. I believe this is a bug.

**Comment by eernstg** [comment]
This issue has been resolved: The CFE chooses `List<int?>` today.
